### ISSUE TITLE
auto_rate - issues with NA in time or oxygen

### ISSUE THREAD

**Issue by nicholascarey** [januarharianto/respR#85]
Just posting this in case it needs to be reproduced. Only time i've seen this in lots of testing with NA values. No idea why it fails. Obviously something to do with how it handles the NA. It works if you set `width` to < 0.18 or > 0.27 !

```
ar_data <- data.frame(time = 1:1000,
                      oxygen = (1 * 2^(seq(1,0,length.out = 1000))))
ar_data_na <- ar_data
ar_data_na[600,2] <- NA

auto_rate(ar_data_na)
## fails with :
# Error in density.default(rollreg$rate_b1, na.rm = T, bw = "SJ-ste", adjust = 0.95) : 
#   need at least 2 points to select a bandwidth automatically 
```

**Comment by nicholascarey** [comment]
Adding to this rather than making new issues, because i've found several issues with `auto_rate` and NA values. It seems in general to handle them poorly, even just a single one, whereas ideally we just want it to ignore them and use the data around them. After all, a lm of several thousand datapoints should not be affected by a single missing value. 

And we do have in the documentation somewhere that respR generally handles NA values well and it shouldn't affect results too much...

So similar to above, get this error (not always though) when there is a single missing value:

```
input_na <- squid.rd
input_na[8001,1] <- NA ##  time NA
auto_rate(input_na)
# Error: `auto_rate(input_na)` threw an error.
# Message: need at least 2 points to select a bandwidth automatically
# Class:   simpleError/error/condition
```

```
input_na <- squid.rd
input_na[8001,2] <- NA ##  oxygen NA
auto_rate(input_na)
# Error: `auto_rate(input_na)` threw an error.
# Message: need at least 2 points to select a bandwidth automatically
# Class:   simpleError/error/condition
```

**Comment by nicholascarey** [comment]
Ok, more.... I wrote this ages and ages ago in a script, so just gonna paste the whole thing here. 

```
## Why does auto_rate give such different results when there is a single NA?
## It shouldn't make much difference. The same lm over that data will not return a radically
## different result just because of one missing datapoint.

## E.g.
urch <- urchins.rd[,1:2]
urch_na <- urch
urch_na[120,2] <- NA

## Very different!
auto_rate(urch)$summary
auto_rate(urch_na)$summary

## I think it's something to do with static_roll, more specifically roll::roll_lm, and how it handles NAs

# Currently if there is even a single NA in either time or oxygen static_roll returns a very different result.
# In fact it loses *width* number of rates entirely

## This is the current static_roll fn

#' Normal rolling regression
#'
#' This is an internal function.
#'
#' @importFrom roll roll_lm
#' @keywords internal
#' @export
static_roll <- function(df, win) {
  roll <- roll_lm(matrix(df[[1]]), matrix(df[[2]]), win)
  roll <- na.omit(data.table(cbind(roll$coefficients, roll$r.squared)))
  setnames(roll, 1:3, c("intercept_b0", "rate_b1", "rsq"))
  return(roll)
}


## This is the one i wrote for use in inspect
## (changed here to roll_reg_plot2 because it's not exported, so run this to use it)

roll_reg_plot2 <- function(df, width) {
  roll_width <- floor(width * nrow(df))
  ## Call all rates, even there is a min_obs of only 1 datapoint
  ## This means rate returned even if there are NA in data
  rates <- roll::roll_lm(matrix(df[[1]]), matrix(df[[2]]),
                         roll_width, min_obs = 1)$coefficients[,2]
  ## However this means rates are ALSO calculated at the start of the data
  ## before the width is even reached, so we remove these.
  rates <- rates[-(1:(roll_width-1))]
  return(rates)
}

## Main difference is the use of min_obs argument in roll_lm.
## This tells it minimum number of observations before it will fit a lm
## Here i have it as 1, so it will fit a lm even if there is only 1 data point
## However, the default for min_obs is the same width as `width` (or `win` in static_roll)
## Which means in static_roll there HAS to be exactly *width* number of datapoints for it to fit the lm.
## So a single NA will mean NO lm calculated for *any* of the rolling regs in which it occurs and
## these are returned as NA.
## So if there are no missing data, roll_lm will return by default a df of the same nrow, but the
## initial rows up to *width* will be NA because there aren't enough datapoints
## But a single NA in the data will mean with static_roll results there will be a *width*
## sized chunk of rows of NAs around it in the output df


## Create some non-linear data
test_data <- data.frame(time = 1:100,
                        oxygen = (1 * 2^(seq(1,0,length.out = 100))))

## inspect - rates decrease as it goes along
inspect(test_data)

## static roll rates
rates_sr <- respR:::static_roll(test_data, floor(0.1*nrow(test_data)))$rate_b1
## roll_reg_plot2 rates
rates_rrp <- roll_reg_plot2(test_data, 0.1)

## plot both
par(mfrow = c(2,1), mar = c(2,2,0.1,0.1))
plot(rates_sr,
     ylim = rev(range(rates_sr)),
     xlim = range(test_data$time))
plot(rates_rrp,
     ylim = rev(range(rates_rrp)),
     xlim = range(test_data$time))

## we can see these are identical
identical(rates_sr, rates_rrp)


## Now what if there is a single missing data point?
test_data_na <- test_data
test_data_na[50,2] <- NA

## static roll rates
rates_sr_na <- static_roll(test_data_na, floor(0.1*nrow(test_data_na)))$rate_b1
## roll_reg_plot2 rates
rates_rrp_na <- roll_reg_plot2(test_data_na, 0.1)

par(mfrow = c(2,1), mar = c(2,2,0.1,0.1))
plot(rates_sr_na,
     ylim = rev(range(rates_sr_na)),
     xlim = range(test_data_na$time))
plot(rates_rrp_na,
     ylim = rev(range(rates_rrp_na)),
     xlim = range(test_data_na$time))

length(rates_sr_na) ## loses *width* number of regressions!
length(rates_rrp_na)

## static roll lose 'width' rate calculations, and because it does na.omit after roll::roll_lm
## any after the NA are shifted backwards by 'width'
## This is what caused problems in the rolling rate plot in inspect, which is why i wrote the
## roll_reg_plot function.
##
## I did add 'min_obs = 1' to static_roll to see what would happen, but it led to lots of errors
## and i don't want to mess with it any more.  I reckon that is the root of the problem. 
## It definitely seems more robust to me to use this. 

## Other questions:
## - Where else - if anywhere - might this cause problems?
## - Is roll_lm used anywhere else with min_obs default?
## - Is there anywhere static_roll outputs are used against other vectors of time/position
## where these rows have *not* been removed? There is no timepoints or location in the output,
## only "intercept_b0", "rate_b1", "rsq"


## So we should fix this, or we have to advise users in inspect that NA data gaps need to be 
## removed or filled.
## Or advise them in auto_rate/calc_rate about NA data that needs to be fixed/removed
## Currently these pass inspect, and i have in the help docs that these problems should not 
## generally cause issues with analysis
## At present, that's clearly wrong...
```

**Comment by nicholascarey** [comment]
Fixed in 5f19afbfdd34a87ddca35efe3b3088bfd3b5f7ce

To summarise: 

roll_lm needs to have `min_obs  =  1` and then it will still return a result if it encounters an NA. But this leaves it with `width` extra lms returned, so these need to be removed from the start. After that, it returns a result of the proper length, even across NA values.
